File: src/test/java/nz/co/searchwellington/filters/LocationParameterFilterTest.java
package nz.co.searchwellington.filters;

import nz.co.searchwellington.geocoding.osm.GeoCodeService;
import nz.co.searchwellington.geocoding.osm.OsmIdParser;
import org.junit.Test;
import org.mockito.Mockito;
import org.springframework.mock.web.MockHttpServletRequest;
import uk.co.eelpieconsulting.common.geo.model.LatLong;
import uk.co.eelpieconsulting.common.geo.model.OsmId;
import uk.co.eelpieconsulting.common.geo.model.OsmType;
import uk.co.eelpieconsulting.common.geo.model.Place;

import static org.junit.Assert.assertEquals;

public class LocationParameterFilterTest {
	
	private static final String VALID_LOCATION = "Petone Station";

	private GeoCodeService geocodeService = Mockito.mock(GeoCodeService.class);

	private Place petoneStation = new Place(VALID_LOCATION, new LatLong(1.1, 2.2), new OsmId(123, OsmType.NODE));

	private LocationParameterFilter filter = new LocationParameterFilter(geocodeService, new OsmIdParser());

	@Test
	public void canResolveOsmIdAsLocation() {
		MockHttpServletRequest request = new MockHttpServletRequest();
		request.setParameter("osm", "123/NODE");
		Mockito.when(geocodeService.resolveOsmId(petoneStation.getOsmId())).thenReturn(petoneStation);
		filter.filter(request);

		final Place locationAttribute = (Place) request.getAttribute(LocationParameterFilter.LOCATION);
		assertEquals(VALID_LOCATION, locationAttribute.getAddress());
	}

	@Test
	public void canResolveLocationSearchRadius() {
		MockHttpServletRequest request = new MockHttpServletRequest();
		request.setParameter("radius", "3");
		filter.filter(request);		
		assertEquals(3.0, request.getAttribute(LocationParameterFilter.RADIUS));
	}
	
	@Test
	public void canResolveAbsoluteLatLongPointAsALocation() {
		MockHttpServletRequest request = new MockHttpServletRequest();
		request.setParameter("latitude", "51.2");
		request.setParameter("longitude", "-0.1");
		
		filter.filter(request);
		
		final Place locationAttribute = (Place) request.getAttribute(LocationParameterFilter.LOCATION);	
		assertEquals(51.2, locationAttribute.getLatLong().getLatitude(), 0);
		assertEquals(-0.1, locationAttribute.getLatLong().getLongitude(), 0);
	}
	
}

File: src/test/scala/nz/co/searchwellington/filters/attributesetters/LocationParameterFilterTest.scala
package nz.co.searchwellington.filters.attributesetters

import nz.co.searchwellington.ReasonableWaits
import nz.co.searchwellington.geocoding.osm.{GeoCodeService, OsmIdParser}
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.mockito.Mockito
import org.mockito.Mockito.when
import org.springframework.mock.web.MockHttpServletRequest
import uk.co.eelpieconsulting.common.geo.model.{LatLong, OsmId, OsmType, Place}

import scala.concurrent.{Await, Future}

class LocationParameterFilterTest extends ReasonableWaits {
  private val VALID_LOCATION_NAME = "Petone Station"
  private val geocodeService = Mockito.mock(classOf[GeoCodeService])
  private val petoneStation = new Place(VALID_LOCATION_NAME, new LatLong(1.1, 2.2), new OsmId(123, OsmType.NODE))

  private val filter = new LocationParameterFilter(geocodeService, new OsmIdParser)

  @Test
  def canResolveOsmIdAsLocation() = {
    val request = new MockHttpServletRequest
    request.setParameter("osm", "123/NODE")
    when(geocodeService.resolveOsmId(petoneStation.getOsmId)).thenReturn(Future.successful(Some(petoneStation)))

    val attributes = Await.result(filter.setAttributes(request), TenSeconds)

    val locationAttribute = attributes("location").asInstanceOf[Place]
    assertEquals(VALID_LOCATION_NAME, locationAttribute.getAddress)
  }

  @Test
  def canResolveLocationSearchRadius() = {
    val request = new MockHttpServletRequest
    request.setParameter("radius", "3")

    val attributes = Await.result(filter.setAttributes(request), TenSeconds)

    assertEquals(3.0, attributes("radius"))
  }

  @Test
  def canResolveAbsoluteLatLongPointAsALocation() = {
    val request = new MockHttpServletRequest
    request.setParameter("latitude", "51.2")
    request.setParameter("longitude", "-0.1")

    val attributes = Await.result(filter.setAttributes(request), TenSeconds)

    val locationAttribute = attributes("location").asInstanceOf[Place]
    assertEquals(51.2, locationAttribute.getLatLong.getLatitude, 0)
    assertEquals(-0.1, locationAttribute.getLatLong.getLongitude, 0)
  }

}